pkg/controller/nodepool/nodepool_controller.go

```diff
@@ -0,0 +1,218 @@
+}
+
+var _ reconcile.Reconciler = &ReconcileNodePool{}
+
+// Add creates a new NodePool Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(c *config.CompletedConfig, mgr manager.Manager) error {
+
+	if !utildiscovery.DiscoverGVK(controllerKind) {
+		klog.Errorf(Format("DiscoverGVK error"))
+		return nil
+	}
+
+	return add(mgr, newReconciler(c, mgr))
+}
+
+type NodePoolRelatedAttributes struct {
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Taints      []corev1.Taint    `json:"taints,omitempty"`
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(c *config.CompletedConfig, mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNodePool{
+		Client:       utilclient.NewClientFromManager(mgr, controllerName),
+		scheme:       mgr.GetScheme(),
+		recorder:     mgr.GetEventRecorderFor(controllerName),
+		Configration: c.ComponentConfig.NodePoolController,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to NodePool
+	err = c.Watch(&source.Kind{Type: &appsv1beta1.NodePool{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Node
+	err = c.Watch(&source.Kind{
+		Type: &corev1.Node{}},
+		&EnqueueNodePoolForNode{})
+	if err != nil {
+		return err
+	}
+
+	npr, ok := r.(*ReconcileNodePool)
+	if !ok {
+		return errors.New(Format("fail to assert interface to NodePoolReconciler"))
+	}
+
+	if npr.Configration.CreateDefaultPool {
+		// register a node controller with the underlying informer of the manager
+		go createDefaultNodePool(mgr.GetClient())
+	}
+	return nil
+}
+
+// +kubebuilder:rbac:groups=apps.openyurt.io,resources=nodepools,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.openyurt.io,resources=nodepools/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reads that state of the cluster for a NodePool object and makes changes based on the state read
+// and what is in the NodePool.Spec
+func (r *ReconcileNodePool) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+
+	// Note !!!!!!!!!!
+	// We strongly recommend use Format() to  encapsulation because Format() can print logs by module
+	// @kadisi
+	klog.Infof(Format("Reconcile NodePool %s/%s", req.Namespace, req.Name))
+
+	ctx := context.Background()
```

Comment:
How about using the context in the `Reconcile` input?

Reply:
good, thanks